### ISSUE TITLE
BUG: backups fail to download to browser

### ISSUE THREAD

**Issue by chris001** [virtualmin/virtualmin-gpl#126]
Found a bug:

When you go to "Backup and Restore", "Backup Virtual Servers", "All virtual servers", "Include sub servers", "Any plan", "Backup all features", "backup destination" = "Download to browser" "Single Archive file", "Halt immediately", "Full (all files)", "Backup Now", after about 5 seconds, it fails with the following error message:
"Error   Backup failed : No features selected to backup"  ?!
That's wrong, "Backup all features" was selected.
1. Why is "Backup virtual servers" broken?
2. Is there any workaround using command line?

Note: 
1. In this server, about 80 GB disk space used on server (that counts Linux OS files obviously), and there's 20 GB free space.  
2. No "cloud storage" available.  
3. Must let Virtualmin create each virtual server's compressed archive backup file ( each one must be < 20 GB in size), then download each backup file thru browser, a few virtual servers per compressed archive backup file, then delete that compressed archive backup file, for this to work within the available 20 GB disk space on server...
4. Yet this fails with the wrong error message, "No features selected to backup".  How to fix it ?!

**Comment by jcameron** [comment]
Odd, I can't re-produce this. Do you get the same error if all the backup settings are the same, but you choose to write to a file on the server instead?

**Comment by chris001** [comment]
When I set it to write to a file on the server, and backup just one virtual server to keep the backup file small, it starts as it should, then it sort of hangs at the TAR step, but it shows the browser notification (lower right corner of screen) that it completed the backup ?!:
```
Starting backup of 1 domains to local file /root/backup ..

Creating backup for virtual server mydomain.com ..

    Copying virtual server configuration ..
    .. done

    Backing up Cron jobs ..
    .. none defined.

    Copying records in DNS domain ..
    .. done

    Saving mail aliases ..
    .. done

    Saving mail and FTP users ..
    .. done

    Backing up mail and FTP user Cron jobs ..
    .. none to backup

    Dumping MySQL database mydomain ..
    .. done

    Dumping MySQL database mydomain_joomla ..
    .. done

    Copying Procmail and SpamAssassin configuration files ..
    .. done

    Backing up Webmin ACL files ..
    .. done

    Creating TAR file of home directory ..
```

**Comment by chris001** [comment]
It wrote out 5 files:
```
-rw-------  1 root root  465836712 Dec  6 16:12 mydomain.com.tar.gz
-rw-r--r--  1 root root       7179 Dec  6 16:12 mydomain.com.tar.gz.dom
-rw-r--r--  1 root root        161 Dec  6 16:12 mydomain.com.tar.gz.info
-rw-------  1 root root      20223 Dec  6 16:12 virtualmin.tar.gz
-rw-r--r--  1 root root        170 Dec  6 16:12 virtualmin.tar.gz.info
```

THE PROBLEM IS THOUGH:  Since I have to do this backup maybe 3-5 parts (estimate 2-3 domains in each part) because of the 20GB disk free space limitation.  
Does each "part" ASSUME that it's the only part restoring to the server, and blows away all previous domains that was previously restored ?
Or does it "additively restore", ie, does NOT destroy all existing domains on the server when you're restoring other domains?

**Comment by jcameron** [comment]
A restore will never delete domains that weren't included in the backup - it will just add new ones, or over-write. 

In the case where the download fails, are you trying to select all domains to include in the backup? Because the backup is written to a temp file on the server first and then sent to the browser, I can't see that working if you have less free space in `/tmp` than the size of the domains.

**Comment by chris001** [comment]
> In the case where the download fails, are you trying to select all domains to include in the backup? Because the backup is written to a temp file on the server first and then sent to the browser, I can't see that working if you have less free space in `/tmp` than the size of the domains.

Yes, all domains were selected, which consume estimated 80GB - Linux Debian 8 (5-10GB) = 70-75GB,.

The error message Virtualmin gives, really should be corrected to be **"Out of disk space while attempting to create backup archive to download in browser".**

**Comment by jcameron** [comment]
Yes, I would expect to get a more useful error message in this case! In fact, I tested it by starting a backup that I knew would fail in this way, but was unable to reproduce the problem you saw..

If you just select one domain to backup, does the same failure happen?

**Comment by chris001** [comment]
> Yes, I would expect to get a more useful error message in this case! In fact, I tested it by starting a backup that I knew would fail in this way, but was unable to reproduce the problem you saw..

Did yours give an error message? Or, did it succeed?  What happened exactly?  Shouldn't it have failed in some way shape or form, when disk space is inadequate to zip and archive of  all the virtual servers on the system?

> If you just select one domain to backup, does the same failure happen?

When I select one domain to backup and download thru the browser, it fails with the error message: `Error  Backup failed : No features selected to backup` EVEN THOUGH all features are selected, having been selected by clicking on the `Select All` button, and also the radio button `Backup all features` is selected.

**Comment by jcameron** [comment]
Yes, when I tried a backup and the tar command failed, I got a useful error message in the browser (not just "No features selected"). Which is what should happen, so clearly there's a bug here..

On your system, when you click the Backup button, do you get a message about "Backup is being prepared, please wait" ?

**Comment by chris001** [comment]
> On your system, when you click the Backup button, do you get a message about "Backup is being prepared, please wait" ?

When you click the `Backup Now` button, a notification appears at the bottom of the page: `Your backup is being prepared. It may take a while. Please wait.` About a second goes by, then the browser makes a popup window (Firefox 72.0b3 implements the popup window as a new tab, and blocks it, and requires you to click on a button to see a dropdown list with the blocked popup and click on the URL of the blocked popup to show it), and the popup tab shows the Virtualmin error message.

**Comment by jcameron** [comment]
Ah .. I wonder if that button to show the popup window is not submitting the original form parameters, which is why you're getting that error. I only tested this in Chrome..

**Comment by chris001** [comment]
> Ah .. I wonder if that button to show the popup window is not submitting the original form parameters, which is why you're getting that error. I only tested this in Chrome..

It appears that indeed the new generation browser security model in firefox most likely does mangle/discard/lose the original form parameters, or the form is nonstandard in some way that inadvertently/unintentionally gets mangled by firefox.

The temporary workaround to get it to work without error, is to go to the huge yellow warning bar at the top of the page, click the button (to view your blocked popups), and then click, `Always allow popups from this site` in the dropdown, to let virtualmin make its old fashioned popups for browser download of virtual server backups.

HUGE DRAWBACK of this feature "download in browser": backing up 444MB took close to 10 minutes of waiting on a black tab, totally uncertain whether the backup would ever complete or ever start downloading as promised.  It did, yet that 10 minutes of nail biting was rather unnecessary... which brings us to this next point:

This "download backup thru browser" feature really should be updated, and it'd be quite basic to implement given today's latest perl modules, to do backups asynchronously in the background, just as cpanel and most other server control panels do them.  For example: You click `Backup Now`, and Virtualmin would say, `Creating this backup zip may take awhile.  Please wait.  Or, Come back to this page later.  When backup is ready to download, a link to download your backup (with virtual server names, time, date, and file size) will be shown at the bottom of this page.`  This way, you could for example get multiple backups queued up in this UI, and Virtualmin would create them in order, one after the other, in the background, and you'd come back after some time, 30 minutes or so, see the download links at the bottom of this `Backup Virtual Servers` page, and click to download each of them to your local storage.

**Comment by chris001** [comment]
It fails to download thru browser, with the message, "Run the backup to local file to view the error message" (!!).
So you run it as a backup to local file, and it fais with: 
```
Starting backup of 1 domains to local file /root/backup ..

Creating backup for virtual server mydomain2.com ..

    Copying virtual server configuration ..
    .. done

    Backing up Cron jobs ..
    .. none defined.

    Copying records in DNS domain ..
    .. done

    Saving mail aliases ..
    .. done

    Saving mail and FTP users ..
    .. done

    Backing up mail and FTP user Cron jobs ..
    .. none to backup

    Copying Webalizer configuration files ..
    .. no logging directive found in webserver configuration!

    Backup failed! See the progress output above for the reason why.
```

How to fix this failure: `Copying Webalizer configuration files .. no logging directive found in webserver configuration!` ?!

**Comment by chris001** [comment]
Does `Backup Virtual Servers` give this `Copying Webalizer configuration files .. no logging directive found in webserver config` error, because the corresponding `nginx` site in `/etc/nginx/sites-enabled` is missing a directive for an `access_log` ie a path to a log file for the particular virtual server?!?

**Comment by jcameron** [comment]
We discussed this feature, and decided that it would be better if the backup was backgrounded and then a link displayed when it was ready to download - that would resolve this whole popup issue. That should come in the next release.

Regarding that "no logging directive found" error, it does indeed mean that an access log is missing from the Nginx config.

**Comment by chris001** [comment]
Related questions:
When backing up to download thru the browser, Virtualmin gives you one file, `backup.tgz`
When backing up to local file system, Virtualmin outputs three files per virtual server: `mydomain.com.tar.gz`, `mydomain.com.tar.gz.dom`, and `mydomain.com.tar.gz.info`, plus optionally two files if you chose to backup virtualmin config: `virtualmin.tar.gz`, and `virtualmin.tar.gz.info`.
1. Are these functionally fully equivalent ?
2. Why not generate the same identical output, no matter whether the destination would be to download to browser, or write to local filesystem?

**Comment by jcameron** [comment]
All the data is in the .tar.gz file, but the .dom and .info files are created when backing up locally to speed up the process of reading backup contents. When downloading, it's currently only possible to send a single file at a time to the browser (without packing them up inside another ZIP file).

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
